tensorflow/contrib/android/java/org/tensorflow/contrib/android/TensorFlowInferenceInterface.java

```diff
@@ -50,6 +56,112 @@
+   * @param  context context The context that contains the files
+   * @param  String filename
+   * @param  boolean res Is it from res/raw
+   *
+   * @return File reference to cache file.
+   */
+  private File copyFile(Context context, String filename, boolean res)
+      throws IOException {
+    InputStream inputStream = null;
+    if (res) {
+      int resourceId = context.getResources().getIdentifier(filename, "raw", context.getPackageName());
+      if (resourceId == 0) {
+        return null;
+      }
+      inputStream = context.getResources().openRawResource(resourceId);
+    } else {
+      inputStream = context.getAssets().open(filename);
+    }
+    File copy = new File(context.getFilesDir(), filename);
+    if (copy.createNewFile() == false) {
+      copy.delete();
+      copy.createNewFile();
+    }
+
+    FileOutputStream fileOutputStream = new FileOutputStream(copy);
+
+    byte[] buffer = new byte[1024];
+    int count;
+    while ((count = inputStream.read(buffer)) != -1) {
+      fileOutputStream.write(buffer, 0, count);
+    }
+    fileOutputStream.close();
+    inputStream.close();
+    return copy;
+  }
+
+  /*
+  * Load a TensorFlow model via FileChannel mmap
+  * This requires that assets files are copied to internal storage already
+  * getFileStreamPath() points to that directory; context used must be passed in here
+  *
+  * @param ctx The context to use to load the model file; same as one used to copy files
+  * @param model The filepath to the GraphDef proto representing the model.
+  */
+  public TensorFlowInferenceInterface(Context ctx, String modelName) {
+    prepareNativeRuntime();
+    File modelFile = null;
+    this.g = new Graph();
+    this.sess = new Session(g);
+    this.runner = sess.runner();
+    this.modelName = modelName;
+    MappedByteBuffer buffer = null;
+    FileChannel fileChannel = null;
+    RandomAccessFile raf = null;
+    File copy = null;
+    try {
+      copy = copyFile(ctx, modelName, true);
+      if (copy != null) {
+        raf = new RandomAccessFile(copy, "r");
+        fileChannel = raf.getChannel();
+      } else {
+        //try assets
+        copy = copyFile(ctx, modelName, false);
+        raf = new RandomAccessFile(copy, "r");
+        fileChannel = raf.getChannel();
+      }
+    } catch (IOException e) {
+      throw new RuntimeException("Failed to open graph file: " + modelFile, e);
+    }
+    try {
+      buffer = fileChannel.map(FileChannel.MapMode.READ_ONLY, 0, fileChannel.size());
+      if (!buffer.isLoaded()) {
+        buffer.load();
+      }
+      byte[] graphDef = new byte[buffer.limit()];
+      buffer.get(graphDef);
+      loadGraph(graphDef, g);
+      Log.i(TAG, "Successfully loaded model from map" + modelName);
+    } catch (IOException ioe) {
+      throw new RuntimeException("Failed to map  model from" + modelName, ioe);
+    } finally {
+      try {
+        if (raf != null) {
+          raf.close();
+        }
+        if (copy != null) {
```

Comment:
Do we want to delete this? What is the copy time -- maybe it would be better to keep it cached if possible for faster loading the next time?